Reduce unnecessary arguments to String.format() in FieldValidator #5369

YongJieYongJie opened the pull request:
Fixes #5369

Turns out have to change `String.format(ERROR_MSG_TEMPLATE, EMAIL_FIELD_NAME, ...)` to something like `ERROR_MSG_TEMPLATE.replace("{fieldName}", EMAIL_FIELD_NAME).replace(...`.

This is because our error message construction mechanism is rather sophisticated, and `String.format()` is not flexible enough (e.g., you cannot insert only the fieldname in one method, and insert the reason for invalidity in another)

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
+
+    // error message components
+    public static final String ERROR_INFO =
+            "\"{userInput}\" is not acceptable to TEAMMATES as {fieldName} because it {reason}.";
```
By the way, the convention for placeholders is `${placeholder}` (missed the dollar sign).

YongJieYongJie commented:
@wkurniawan07 Ready for first review!

Summary of changes:
 - changed `String.format()` to `String.replace()`
  - i.e., from using `%s` to using `${placeholders}`
 - "generalized" field-specific error messages
   - e.g., from `"...not acceptable as " + SOME_FIELD_NAME + " because..."` to `"...not acceptable as ${fieldName} because..."`
   - this allowed me to remove many of the redundant fields, as in f29fa2f
 - re-ordered and tidied FieldValidator in general
  - (and since I was moving code around, I kept renaming of variables to the bare minimal so it's easier to understand the changes)
 - finally, I realized using chains of `String.replace()` is too verbose, and thus created a helper method `getInterpolatedErrorMessage()`

Note / things left for other issues:
 - other tests using error messages in their assertion should import only the required error messages and not the whole of `FieldValidator`, this will make the code less verbose because there is no need to prepend the constant with `FieldValidator.`
  - in any case, perhaps such tests should not assert the existence of _specific_ error messages but to just check that there was indeed an error; this is because there are html checks and FieldValidatorTest

 - I also changed _some_ expected error messages in `FieldValidatorTest` into string literals
  - previously, these error messages were constructed using `String.format()`,
  - however, using `String.format()` to generate test expectations is wrong because we are ultimately testing that our `String.format()` are working fine;
  - __also, there should be a place where we can see what all the final error messages look like__
  - nonetheless, this changing from `String.format()` to string literals is only partially done because I realized that there are quite some test redundancies and it is better done in a separate issue


 - I have mainly dealt with the "name-related" messages which were very similar to `SIZE_CAPPED_NON_EMPTY_STRING_ERROR_MESSAGE`
  - other error messages are quite specific in their area of application and can be refactored separately (e.g., `PARTICIPANT_TYPE_ERROR_MESSAGE`, `GENDER_ERROR_MESSAGE`, etc.)

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -13,77 +13,47 @@
+    /////////////////
+
+    // name-related
+    public static final String PERSON_NAME_FIELD_NAME = "a person name";
```
Sometimes the field name has the article "a/an", sometimes it doesn't. Let's aim to fix this one as well while we're at it.

YongJieYongJie replied:
@wkurniawan07 Fixing this will require modifying some of the final error messages, and I'm actually leaving wording changes for another issue. (I forgot to mention this in the pull request description.)

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
      * problematic if we allow multiple instructors for a single course.
      * TODO: make case insensitive
      */
+    public static final String COURSE_ID_FIELD_NAME = "a Course ID";
```
I don't think the letter "[C]ourse" needs to be capitalised.

YongJieYongJie replied:
As per above, I'm trying not to make any changes in wording in this PR.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
-     */
-    public static final String START_TIME_FIELD_NAME = "start time";
-    public static final String END_TIME_FIELD_NAME = "end time";
+    // TODO: allow smaller increments (currently only 1 hour increments allowed)
```
Not your fault, but this TODO has nothing to do with the field name. Let's move it to the more relevant place.

YongJieYongJie replied:
I'll create an issue for it

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
+    
+    public static final String GIVER_TYPE_NAME = "feedback giver.";
+    public static final String RECIPIENT_TYPE_NAME = "feedback recipient.";
+    public static final String VIEWER_TYPE_NAME = "feedback viewer.";
```
Why do these need to end with a full stop?

YongJieYongJie replied:
It's because the `PARTICIPANT_TYPE_ERROR_MESSAGE` from which the full error message is constructed does not has the full stop. I'm leaving this error message for a separate issue.

I am dealing mainly with error messages related to `SIZE_CAPPED_NON_EMPTY_STRING_ERROR_MESSAGE` in this PR. Previously, almost the whole of `FieldValidator` was rather confusing; this was because `SIZE_CAPPED...` constituted a large part of `FieldValidator` and was itself rather confusing. My hope is that after this PR, `FieldValidator` looks like something with many separate mistakes (to be fixed separately), and not one big mistake.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
+    public static final String HINT_FOR_CORRECT_FORMAT_FOR_SIZE_CAPPED_NON_EMPTY =
+            "The value of ${fieldName} should be no longer than ${maxLength} characters. It should not be empty.";
+    public static final String HINT_FOR_CORRECT_FORMAT_FOR_SIZE_CAPPED_POSSIBLY_EMPTY =
+            "The value of ${fieldName} should be no longer than ${maxLength} characters.";
```
The non-empty version is basically the possibly empty version plus "It should not be empty.". Let's not repeat ourselves here.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
+            "Trying to update to an email that is already used by: %s/%s";
+
+    public static final String HINT_FOR_CORRECT_COURSE_ID =
+            "A Course ID can contain letters, numbers, fullstops, hyphens, underscores, and dollar signs. "
```
Similarly here for "[C]ourse".

YongJieYongJie replied:
Noted. But as per above, I'm trying not to make any changes in wording in this PR.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
+    public static final String HINT_FOR_CORRECT_EMAIL =
+            "An email address contains some text followed by one '@' sign followed by some more text. "
+            + "It cannot be longer than ${maxLength} characters. "
+            + "It cannot be empty and it cannot have spaces.";
```
These two sentences can be combined: "It cannot be longer than ${maxLength} characters, cannot be empty, and cannot contain spaces."

YongJieYongJie replied:
Noted. But as per above, I'm trying not to make any changes in wording in this PR.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
+            + "It cannot be longer than ${maxLength} characters. "
+            + "It cannot be empty or contain spaces.";
+    public static final String COURSE_ID_ERROR_MESSAGE =
+            ERROR_INFO + " " + HINT_FOR_CORRECT_COURSE_ID;
```
Similarly here.

YongJieYongJie replied:
Noted. But as per above, I'm trying not to make any changes in wording in this PR.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
+    public static final String EMAIL_ERROR_MESSAGE =
+            ERROR_INFO + " " + HINT_FOR_CORRECT_EMAIL;
+    public static final String EMAIL_TAKEN_MESSAGE =
+            "Trying to update to an email that is already used by: %s/%s";
```
Not your fault, but an email being taken has nothing to do with field validation. See if you can move this constant out of here.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -103,149 +73,107 @@
+    public static final String HINT_FOR_CORRECT_FORMAT_OF_GOOGLE_ID =
+            "A Google ID must be a valid id already registered with Google. "
+            + "It cannot be longer than " + GOOGLE_ID_MAX_LENGTH + " characters. "
+            + "It cannot be empty.";
```
Similarly here (combine the sentences).

YongJieYongJie replied:
Noted. But as per above, I'm trying not to make any changes in wording in this PR.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -366,13 +263,17 @@ public String getInvalidityInfoForEmail(String email) {
         if (email.isEmpty()) {
-            return String.format(EMAIL_ERROR_MESSAGE, email, REASON_EMPTY);
+            return getInterpolatedErrorMessage(
+                    EMAIL_ERROR_MESSAGE, email, EMAIL_FIELD_NAME, REASON_EMPTY, EMAIL_MAX_LENGTH);
```
Here (and in many parts after here) it should be fine to include the parameters in the same line since it's not squished at all:
```java
return getInterpolatedErrorMessage(EMAIL_ERROR_MESSAGE, email, ...,
                                   ...);
```

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -813,4 +726,22 @@ private boolean isUntrimmed(String value) {
+    /////////////////////////////////////
+    // HELPER METHODS FOR INTERNAL USE //////////////////////////
+    // DUPLICATED IN FieldValidatorTest FOR EXTERNAL TESTS USE //
+    /////////////////////////////////////////////////////////////
```
Don't need to include this part. Rather, just add header comments to the following two functions.
Also, consider `getPopulated*` instead of `getInterpolated*`.

wkurniawan07 replied:
In fact, just make the following two methods public. As a utility class, this class can be called by anyone, but a test case should not be dependent on another, non-base test case, i.e `test.cases.*` -> `test.cases.*` is a dependency that should be avoided.

YongJieYongJie replied:
@wkurniawan07 I think the methods should not be public because other (non-test) classes should not be constructing error messages themselves, but should rather use the specific `getInvalidityInfoFor...()` provided.

I'll remove the comments and the duplicated methods in `FieldValidatorTest`

wkurniawan07 commented on src/test/java/teammates/test/cases/automated/SubmissionsAdjustmentTest.java:
```diff
@@ -32,6 +32,7 @@
 import teammates.logic.core.FeedbackResponsesLogic;
 import teammates.logic.core.FeedbackSessionsLogic;
 import teammates.logic.core.StudentsLogic;
+import teammates.test.cases.common.FieldValidatorTest;
```
Doesn't make sense for a test case to be dependent on another (I specifically target this in #5570).

wkurniawan07 commented on src/test/java/teammates/test/cases/storage/InstructorsDbTest.java:
```diff
@@ -320,10 +327,14 @@ public void testUpdateInstructorByGoogleId() throws Exception {
+                    + FieldValidatorTest.getInterpolatedErrorMessage(
+                          FieldValidator.EMAIL_ERROR_MESSAGE, instructorToEdit.email,
+                          FieldValidator.EMAIL_FIELD_NAME, FieldValidator.REASON_INCORRECT_FORMAT,
+                          FieldValidator.EMAIL_MAX_LENGTH),
```
The indentation here is inconsistent with the one below.

wkurniawan07 commented:
@YongJieYongJie good job in general! The class is much neater than before. There are quite a lot of rooms for improvement to make it even neater.

YongJieYongJie commented:
@wkurniawan07 Ready for re-review!

I kept the helper methods `getInterpolatedErrorMessage()` in `FieldValidator` private because other classes should not be constructing their own error messages.

I removed the helper methods in `FieldValidatorTest` to remove the dependency between tests.

Also, I forgot to mention earlier that I'm trying not to make any change to the actual error messages because that is a big issue by itself. Nonetheless, I standardized the `a/an`, combined the `not empty/no spaces`, changed `[C]ourse` to lowercase, and removed the duplicated full-stop.

wkurniawan07 commented on src/main/java/teammates/common/util/FieldValidator.java:
```diff
@@ -487,8 +390,8 @@ public String getInvalidityInfoForGender(String gender) {
-     * {@code COURSE_NAME_MAX_LENGTH}, does not contain any invalid characters (| or %), and has no unsantized
-     * HTML characters
+     * {@code FEEDBACK_SESSION_NAME_MAX_LENGTH}, does not contain any invalid characters (| or %), and has no
+     * unsantized HTML characters
```
Let's fix the typo as well: `unsanitized`

YongJieYongJie commented:
@wkurniawan07 Are there other change requests?

wkurniawan07 commented on src/test/java/teammates/test/cases/automated/SubmissionsAdjustmentTest.java:
```diff
@@ -218,8 +218,11 @@ public void testEnrollStudentsWithScheduledSubmissionAdjustment() throws Excepti
+                                     .replace("${userInput}", "e6@g@")
+                                     .replace("${fieldName}", FieldValidator.EMAIL_FIELD_NAME)
+                                     .replace("${reason}", FieldValidator.REASON_INCORRECT_FORMAT)
+                                     .replace("${maxLength}", String.valueOf(FieldValidator.EMAIL_MAX_LENGTH));
```
Granted that the reason for keeping the `getPopulatedErrorMessage` is valid, I don't really like repetitions of this procedure. Let's use reflection here instead. Add this method to `BaseTestCase.java`:
```java
protected static String getPopulatedErrorMessage(String messageTemplate, String userInput, String fieldName,
                                                 String errorReason, int maxLength) {
    try {
        Method getPopulatedErrorMessage = FieldValidator.class.getDeclaredMethod("getPopulatedErrorMessage",
                String.class, String.class, String.class, String.class, int.class);
        getPopulatedErrorMessage.setAccessible(true);
        return (String) getPopulatedErrorMessage.invoke(null, messageTemplate, userInput, fieldName,
                                                        errorReason, maxLength);
    } catch (Exception e) {
        throw new RuntimeException(e);
    }
}
```
And you can call `getPopulatedErrorMessage` from any test case without breaking the abstraction. Do the same with the 4-params version.

wkurniawan07 commented:
@YongJieYongJie other than the two suggestions, can you clarify again why a small subset of the error messages are hardcoded, and why those subset instead of the others?

YongJieYongJie commented:
@wkurniawan07 The plan was to hardcore all the error messages in `FieldValidatorTest`. Reason being to create this one place where we can see the full error messages and easily detect possible mistakes when using `String.format()`/`String.replace()` (e.g., missing spaces etc.).

`FieldValidatorTest` will then also work sort of like a spec file: when you want to rephrase the error messages, you go in and change the error message there, run the tests and watch them fail, fix the tests...

However, because there is currently quite a bit of redundant test cases in `FieldValidatorTest`, changing all test expectations to hardcoded error message will clog up the file. As such, I am intending to refactor the tests separately and introduce the hardcoded error messages then.

As for the inconsistent state (some messages are hardcoded while others are constructed) in other test files, it is because I was halfway through changing them from hardcoded error messages to using `String.format()`, then @damithc noted that having hardcoded error messages is useful (for reasons mentioned above). So I put that pull request on hold (#5360), and is trying to localize the hardcoded error messages to `FieldValidatorTest`

YongJieYongJie commented:
@wkurniawan07 Added the `getPopulatedErrorMessage()` to `BaseTestCase.java`, and also changed all tests expectations to be hardcoded.

I now realized why the changes seemed to have made the code more verbose:
 - previously, the "name-related" fields (e.g., person name, nationality, etc.) have field-specific error messages so we don't have to pass in the `fieldName` and `maxLength` etc.
 - however I generalized these specific error messages because they are misleading: because such fields have both specific and general messages, and each with different number of `%s` needed for the construction
 - (e.g., even though there is a `PERSON_NAME_ERROR_MESSAGE`, depending on the reason of invalidity, sometimes an invalid person name may instead generate stuff like `WHITESPACE_ONLY_OR_EXTRA_WHITESPACE_ERROR_MESSAGE`, which has a different "hint" component and requires passing in of the `fieldName` and `maxLength`)

wkurniawan07 commented on src/test/java/teammates/test/cases/BaseTestCase.java:
```diff
@@ -114,6 +116,25 @@ protected static void ignorePossibleException() {
+                            String.class, String.class, String.class, int.class);
+            getPopulatedErrorMessage.setAccessible(true);
+            return (String) getPopulatedErrorMessage.invoke(null, messageTemplate, userInput, fieldName,
+                                                            errorReason, maxLength);
```
Anyway, I understand your reasoning for not wanting to invoke the private method. There is some validity to that. If you wish so, you can change this method to copy the definition of `getPopulatedErrorMessage` in `FieldValidator`. My concern when you did that in `FieldValidatorTest` was that it caused strange dependency to `FieldValidatorTest`, but if the method is in `BaseTestCase` like right now, that is an acceptable dependency.

YongJieYongJie replied:
@wkurniawan07  Yup yup, it is perfectly valid if this helper is provided in the `BaseTestCase`. I totally misread your previous comment and was under the impression that you suggested using reflection directly in each test =(

wkurniawan07 commented on src/test/java/teammates/test/cases/common/FieldValidatorTest.java:
```diff
@@ -193,38 +198,43 @@ public void testGetValidityInfoForAllowedName() {
+                    "\"Dr. Amy-Bén s&#x2f;o O&#39;&amp;|% 2\t\n (~!@#$^*+_={}[]\\:;&quot;&lt;&gt;?)\" is not "
+                    + "acceptable to TEAMMATES as a/an name field because it contains invalid characters. All "
+                    + "name field must start with an alphanumeric character, and cannot contain any vertical "
+                    + "bar (|) or percent sign (%).",
```
Looks like this is off by one space :p

wkurniawan07 commented on src/test/java/teammates/test/cases/common/FieldValidatorTest.java:
```diff
@@ -55,9 +55,9 @@ public void testGetValidityInfoForSizeCappedNonEmptyString() {
-                        tooLongName, typicalFieldName, REASON_TOO_LONG, typicalFieldName, maxLength),
+                "\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\" is not acceptable to TEAMMATES as "
+                + "a/an my field because it is too long. The value of a/an my field should be no longer than 50 "
+                + "characters. It should not be empty.",
```
The ones that you just moved here (from being defined as constants) don't have the line break indentations. Don't forget to add them.

wkurniawan07 commented:
Almost there, just a few suggestions to consider!

wkurniawan07 commented on src/test/java/teammates/test/cases/common/FieldValidatorTest.java:
```diff
@@ -250,51 +250,51 @@ public void testGetValidityInfoForAllowedName() {
+                     "\"\" is not acceptable to TEAMMATES as a/an name field because it is empty. The value "
+                         + "of a/an name field should be no longer than 50 characters. It should not be empty.",
+                     validator.getValidityInfoForAllowedName(typicalFieldName, maxLength,
+                                                             emptyValue));
```
This one as well.

wkurniawan07 commented on src/test/java/teammates/test/cases/common/FieldValidatorTest.java:
```diff
@@ -250,51 +250,51 @@ public void testGetValidityInfoForAllowedName() {
+                     "The provided name field is not acceptable to TEAMMATES as it contains only whitespace "
+                         + "or contains extra spaces at the beginning or at the end of the text.",
+                     validator.getValidityInfoForAllowedName(typicalFieldName, maxLength,
+                                                             untrimmedValue));
```
This one as well.

wkurniawan07 commented on src/test/java/teammates/test/cases/common/FieldValidatorTest.java:
```diff
@@ -480,37 +503,55 @@ public void testGetInvalidityInfoForEmail_valid_returnEmptyString() {
+                     "\"\" is not acceptable to TEAMMATES as a/an email because it is empty. An email "
+                         + "address contains some text followed by one '@' sign followed by some more text. "
+                         + "It cannot be longer than 254 characters, cannot be empty and cannot contain "
+                         + "spaces.",
```
This one as well.

wkurniawan07 commented on src/test/java/teammates/test/cases/common/FieldValidatorTest.java:
```diff
@@ -303,26 +303,32 @@ public void testGetInvalidityInfoForNationality_invalid_returnSpecificErrorStrin
+                     "\"Percent Symbol % Section\" is not acceptable to TEAMMATES as a/an section name "
+                         + "because it contains invalid characters. All section name must start with an "
+                         + "alphanumeric character, and cannot contain any vertical bar (|) or percent sign "
+                         + "(%).",
```
This one as well.

wkurniawan07 commented on src/test/java/teammates/test/cases/common/FieldValidatorTest.java:
```diff
@@ -250,51 +250,51 @@ public void testGetValidityInfoForAllowedName() {
+                     "\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\" is not "
+                         + "acceptable to TEAMMATES as a/an institute name because it is too long. The value "
+                         + "of a/an institute name should be no longer than 64 characters. It should not be "
+                         + "empty.",
```
This one and the one above (both instances of `empty.`) can also go to the line before.

wkurniawan07 commented:
Some cosmetic changes requested, which should be the last from me!
